Use pre-commit to set license headers

joaander opened the pull request:
<!-- Please confirm that your work is based on the correct branch. -->
<!-- Bug fixes should based on *maint*. -->
<!-- New features should based on *master*. -->

## Description

<!-- Describe your changes in detail. -->
Add a Python script run in pre-commit to ensure that all files have a proper license header.

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->
Reduce maintainer burden in updating these. With this in place, all files will always have a proper header and the copyright date needs to only be updated in one place.

## How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->
I ran pre-commit locally and observed the diff.

<!--- Please build the sphinx documentation and check that any changes to
      documentation display properly. -->

## Change log

No change log entry needed.

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/gsd/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**GSD Contributor Agreement**](https://github.com/glotzerlab/gsd/blob/master/ContributorAgreement.md).
- [x] My name is on the list of contributors (`doc/credits.rst`) in the pull request source branch.

joaander commented:
@bdice @b-butler Have a look at this script for use with in pre-commit. I'm testing it out on GSD first as it is a smaller project. I hope to move it to HOOMD to ensure that all files have a license header and to reduce the effort needed to maintain those headers.

bdice commented on .github/workflows/fix-license-header.py:
```diff
@@ -0,0 +1,95 @@
+
+            status = fix_file(f=f,
+                              header_lines=header_lines,
+                              prefix=COMMENT_PREFIX[extension])
```
Would you want to check if the extension is supported and throw a user-friendly error messasge?

joaander replied:
Yes, if we keep the comment prefixes in this file. If we make it a command line arg (https://github.com/glotzerlab/gsd/pull/105#issuecomment-846160594) this won't be needed.

bdice reviewed (approved):
This looks great. Would it be possible to make this a shared pre-commit hook that we can use for all our projects?

bdice commented on CHANGELOG.rst:
```diff
@@ -1,3 +1,5 @@
+.. Copyright (c) 2016-2021 The Regents of the University of Michigan
+.. This file is part of GSD, released under the BSD 2-Clause License.
```
I think these files need a newline after the `..` comments.
```suggestion
.. This file is part of GSD, released under the BSD 2-Clause License.

```

joaander replied:
Yes, I should teach the script to write a blank line after the header if it is not yet present.

joaander commented:
> This looks great. Would it be possible to make this a shared pre-commit hook that we can use for all our projects?
> 
> Also this is going to be a huge help for contributors creating new files -- people often forget to add this kind of info at the top.

Yes, this is just a prototype. I do plan to make it reusable. I looked at the existing pre-commit hooks and nothing exactly matched what I was looking for. There was only one that was close, but it was mainly a validator and required a convoluted process to fix files: https://github.com/Lucas-C/pre-commit-hooks

For a reusable script, should we could embed comment prefixes in it for all the types we are interested in? Or should we make it a command line argument? The latter is more flexible, but at the cost of additional pre-commit config lines:
```
  - id: fix-license-header
    types_or: [python, cython]
    args:
    - --comment='#'
  - id: fix-license-header
    types_or: [c, c++, cuda]
    args:
    - --comment='//'
  - id: fix-license-header
    types_or: [rst]
    args:
    - --comment='..'
```

b-butler commented on doc/changes.rst:
```diff
@@ -1 +1,3 @@
-.. include:: ../CHANGELOG.rst
```
These seems to have deleted the include directive without replacing it at the bottom. My guess since rst uses the same prefix for a directive as a comment, this thought it was the incorrect comment, not a directive.

b-butler commented on doc/conf.py:
```diff
@@ -1,5 +1,7 @@
 #!/usr/bin/env python3
-# -*- coding: utf-8 -*-
```
I am fine with not including the encoding, but this was deleted without replacement again.

b-butler commented on doc/installation.rst:
```diff
@@ -1 +1,3 @@
-.. include:: ../INSTALLING.rst
```
Removal of a directive again.

b-butler reviewed (changes requested):
I really like this idea, especially because I often forget to do this.

Is there a reason we are opening the files in read plus binary mode? In text mode we do not have to deal with line endings `\n` will work for all platforms according to Python's documentation, and we don't seem to be escaping much or doing much with binary mode?

Also some languages and packages use comments/comment prefixes for various meanings e.g. Python using it for character encodings and RST directives. The RST seems to be the bigger problem since a file with a single include will just become a file with just a license message. To fix this would necessitate making the script more complicated, but would help prevent accidental errors like this more often.

joaander commented:
> I really like this idea, especially because I often forget to do this.
> 
> Is there a reason we are opening the files in read plus binary mode? In text mode we do not have to deal with line endings `\n` will work for all platforms according to Python's documentation, and we don't seem to be escaping much or doing much with binary mode?

I followed the same process as in `pre-commit-hooks` https://github.com/pre-commit/pre-commit-hooks/blob/master/pre_commit_hooks/fix_encoding_pragma.py which uses the binary mode. I think it does this to preserve the line endings used in the file and avoid generating files with mixed line endings or otherwise rewriting all the line endings in a file when adding a header.

> Also some languages and packages use comments/comment prefixes for various meanings e.g. Python using it for character encodings and RST directives. The RST seems to be the bigger problem since a file with a single include will just become a file with just a license message. To fix this would necessitate making the script more complicated, but would help prevent accidental errors like this more often.

The difficulty is how to detect whether a given file line should be replaced or not. If we never replace lines, then we will end up with multiple headers when we change the copyright year or otherwise reword the header. If we make this a checker only, then it isn't as useful. As I move this to its own command line tool, I'll add a 'keep' option so that callers can add specific line signatures to keep in their pre-commit configs. That moves the complexity to the caller so that we don't have to support all exceptions built in in the code.

joaander commented:
I updated this working with the new reusable fixer https://github.com/glotzerlab/fix-license-header. @b-butler I addressed your concerns with `--keep-before` and `--keep-after` settings.

Compared to the last review, this PR is forced pushed with a clean update of the pre-commit config and then a run of pre-commit.

I'm going to wait a bit on tagging a 1.0 release of https://github.com/glotzerlab/fix-license-header until we try this on a few more tools to ensure that it has the options we need.

b-butler reviewed (approved):
LGTM
